L81-aspect-based-python-bazel-rules.md

```diff
@@ -0,0 +1,50 @@
+Aspect-based Python Bazel Rules
+----
+* Author(s): Michael Beardsworth
+* Approver: a11r
+* Status: Draft
+* Implemented in: Bazel and Starlark
+* Last updated: [Date]
+* Discussion at: <google group thread> (filled after thread exists)
+
+## Abstract
+
+Proposes modifications to gRPC's Bazel rules for Python codegen to improve
+dependency tracking.
+
+## Background
+
+* gRPC provides build rule logic written in Starlark to support Protobuf and gRPC Python codegen with Bazel.
+* These rules (`py_proto_library` and `py_grpc_library`) each generate code from a single `proto_library`'s protos.
+* The produced Python libraries (or equivalently-functioning PyInfo providers) do not express dependencies on the generated Python code for the `proto_library`'s dependencies.
+* Users are surprised by this behavior. See [#23769](https://github.com/grpc/grpc/issues/23769) for an example bug.
+* This behavior is particularly surprising for Google-internal users. The internal versions of `py_proto_library` and `py_grpc_library` propagate Python dependencies.
+
+### Related Proposals:
+`py_proto_library` and `py_grpc_library` are implemented in [python_rules.bzl](https://github.com/grpc/grpc/blob/bazel/python_rules.bzl). There doesn't seem to be a proposal for that, though.
```

Comment:
> There doesn't seem to be a proposal for that, though.

To give you some background here, these two rules were originally intended to be used only by the `grpc/grpc` repo. It wasn't until later that we found that lots of folks were using them externally as well.

Reply:
Hyrum's Law in action! It certainly seems like there's demand - maybe it's worth considering upstreaming this to [https://github.com/bazelbuild/rules_python](https://github.com/bazelbuild/rules_python) in a future change.

Reply:
Perhaps. Several other repos (including I think `rules_python`) had similar rules but they were much less rosy about other repos using their rules. There would definitely need to be a discussion about it.